compiler/rustc_const_eval/src/interpret/place.rs

```diff
@@ -641,12 +641,18 @@ where
         let layout_compat =
             mir_assign_valid_types(*self.tcx, self.param_env, src.layout, dest.layout);
         if !allow_transmute && !layout_compat {
-            span_bug!(
-                self.cur_span(),
-                "type mismatch when copying!\nsrc: {:?},\ndest: {:?}",
-                src.layout.ty,
-                dest.layout.ty,
-            );
+            if let Err(e) = src.layout.ty.error_reported() {
+                throw_inval!(AlreadyReported(e.into()))
+            } else if let Err(e) = dest.layout.ty.error_reported() {
+                throw_inval!(AlreadyReported(e.into()))
+            } else {
+                span_bug!(
+                    self.cur_span(),
+                    "type mismatch when copying!\nsrc: {:?},\ndest: {:?}",
+                    src.layout.ty,
+                    dest.layout.ty,
+                );
+            }
```

Comment:
We shouldn't even run the interpreter on such code, so this looks to me like there is a bug elsewhere in rustc?

We certainly don't want to sprinkle the entire interpreter with checks like this -- there is no reason for me to believe that doing such a check only here in `copy_op_no_validate` is sufficient.

Cc @rust-lang/wg-const-eval

Reply:
I don't know that there's a bug elsewhere, rather that the compiler is just trying to do as much as it can before bailing out

Reply:
Yeah but it is being too aggressive at that. :) This is similar to https://github.com/rust-lang/rust/issues/79047 and https://github.com/rust-lang/rust/issues/97477: consts should not get evaluated if they refer to "tainted" things. We already do that for many cases; in https://github.com/rust-lang/rust/issues/112824 a pattern is involved which presumably circumvents the guards we have on place to avoid this situation for regular const-eval.

In fact it seems like a pattern and a RPIT is needed to hit this issue? [This variant](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=926e047a528f3e0a295113d394db7557) works fine. I hope whatever prevents `msg_type` from being evaluated there, can also be used to stop evaluating it when RPIT are present. Cc @rust-lang/types

Reply:
The query stack should elucidate why a pattern + an RPIT is necessary -- validating the RPIT forces us to call `type_of` -> `borrowck` -> `mir_built` (on the RPIT's parent function) -> match exhaustiveness checking code -> `eval_to_allocation_raw` (on the const) -> generating MIR for the const -> `ConstProp`.

Not exactly sure if it's *feasible* to bubble up the error in order to taint things, but that would be the ideal fix for this, yeah.

Reply:
To find the site that creates the ty::Error you should be able to use `-Ztreat-err-as-bug` (potentially with some integers > 1 to find the right one)

Reply:
> To find the site that creates the ty::Error you should be able to use `-Ztreat-err-as-bug` (potentially with some integers > 1 to find the right one)

`-Ztreat-err-as-bug` wasn't helping, both of the earlier errors come from `rustc_resolve` (`rustc_resolve/src/late/diagnostics.rs:2242:9` and `rustc_resolve/src/diagnostics.rs:176:17`), neither of which are where the `ty::Error` is created.

Reply:
you can use higher indices than the number of errors to try to find the place where `delay_span_bug` happens (but I've seen that be buggy before, so ymmv).

it's also odd. I would have imagined this to be an astconv issue where we go from a `Res::Err` to a `ty::Error`, but the site where we do that actually taints.

Maybe it's `ast_ty_to_ty`? it does `hir::TyKind::Err(guar) => tcx.ty_error(*guar),`, so you may need to taint first.

Reply:
> you can use higher indices than the number of errors to try to find the place where delay_span_bug happens (but I've seen that be buggy before, so ymmv).

I didn't know this - neat!

> it's also odd. I would have imagined this to be an astconv issue where we go from a Res::Err to a ty::Error, but the site where we do that actually taints.

So we actually see a bunch of delayed bugs here:

- `rustc_middle/src/ty/sty.rs:1540` which creates a `Region::new_error`, no tainting.
- `rustc_hir_analysis/src/astconv/mod.rs:2724` which is where you've mentioned above, tainting happens here but doesn't seem to ultimately make a difference.
- `rustc_borrowck/src/type_check/mod.rs:280` (from `mod.rs:2448`)
- ..previous one again.

I'll look into why we are in fact tainting but it doesn't seem to matter.

Reply:
is that during typeck of the constant? or something else?

Reply:
I did some more digging and chatted to @oli-obk and here's what we've worked out so far: `set_tainted_by_errors` does get invoked earlier due to astconv - as expected - however, this is in `ItemCtxt`'s `AstConv` implementation, which is a no-op, so the tainting doesn't actually do anything. We'll need to work out a way to keep track of this and hopefully that will be the principled way to resolve this.